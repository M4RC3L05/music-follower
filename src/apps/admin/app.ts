import config from "config";
import Koa from "koa";
import bodyParser from "koa-bodyparser";
import render from "koa-ejs";
import qs from "koa-qs";
import session from "koa-session";
import koaStatic from "koa-static";
import path from "node:path";
import url from "node:url";

import router from "#src/apps/admin/router.js";

export const app = () => {
  const koa = new Koa();
  koa.keys = config.get("apps.admin.keys");

  qs(koa);
  render(koa, {
    root: path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), "views"),
    layout: "main",
    viewExt: "ejs",
    cache: process.env.NODE_ENV === "production",
    debug: false,
  });

  koa.use(session({ ...config.get("apps.admin.session") }, koa));
  koa.use(bodyParser());
  koa.use(koaStatic(path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), "../../../public")));
  koa.use(router.middleware());
  koa.use(router.allowedMethods());

  return koa;
};
