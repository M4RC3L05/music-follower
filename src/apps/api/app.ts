import { type Database } from "@leafac/sqlite";
import config from "config";
import { Hono } from "hono";
import { basicAuth } from "hono/basic-auth";
import { cors } from "hono/cors";
import { errorMappers } from "#src/errors/mod.js";
import { errorMapper, requestLifeCycle } from "#src/middlewares/mod.js";
import { router } from "./router.js";

export type Api = ReturnType<typeof makeApp>;

export const makeApp = ({ database }: { database: Database }) => {
  const app = new Hono();

  app.use("*", (c, next) => {
    c.set("database", database);

    return next();
  });
  app.use("*", requestLifeCycle);
  app.use("*", cors());
  app.use(
    "*",
    basicAuth({
      username: config.get<{ name: string; pass: string }>("apps.api.basicAuth")
        .name,
      password: config.get<{ name: string; pass: string }>("apps.api.basicAuth")
        .pass,
    }),
  );

  app.onError(
    errorMapper({
      mappers: [errorMappers.validationErrorMapper],
      defaultMapper: errorMappers.defaultErrorMapper,
    }),
  );

  return app.route("/api", router());
};
