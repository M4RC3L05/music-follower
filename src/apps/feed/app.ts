import { App } from "@m4rc3l05/sss";

import { feedMiddleware } from "./middlewares/feed-middleware.js";
import requestLifeCycle from "#src/middlewares/request-lifecycle.js";

export const makeApp = () => {
  const app = new App();

  app.use(requestLifeCycle);
  app.use(feedMiddleware);

  return app;
};
