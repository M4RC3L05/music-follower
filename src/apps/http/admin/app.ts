import Koa from "koa";
import basicAuth from "koa-basic-auth";
import config from "config";
import koaStatic from "koa-static";

export const app = () => {
  const koa = new Koa();

  koa.use(basicAuth({ ...config.get("apps.admin.basicAuth") }));
  koa.use(koaStatic("./src/apps/http/admin/public"));

  return koa;
};
