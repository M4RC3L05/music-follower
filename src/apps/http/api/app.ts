import Koa from "koa";
import bodyParser from "koa-bodyparser";
import cors from "@koa/cors";

import * as errorMappers from "#src/common/errors/mappers/mod.js";
import { errorMapper, requestLifeCycle } from "#src/common/middlewares/mod.js";
import logger from "#src/common/clients/logger.js";
import router from "./router.js";

export const app = () => {
  const app = new Koa();

  app.use(requestLifeCycle({ loggerFactory: logger }));
  app.use(
    errorMapper({
      loggerFactory: logger,
      defaultMapper: errorMappers.defaultErrorMapper,
      mappers: [errorMappers.validationErrorMapper],
    }),
  );
  app.use(cors());
  // App.use(basicAuth({ ...config.get("apps.api.basicAuth") }));
  app.use(bodyParser());
  app.use(router.routes());
  app.use(router.allowedMethods());

  return app;
};
