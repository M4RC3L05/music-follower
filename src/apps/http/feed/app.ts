import Koa from "koa";
import favicon from "koa-favicon";

import { feedMiddleware } from "#src/apps/http/feed/middlewares/feed-middleware.js";

export const app = () => {
  const koa = new Koa();

  koa.use(favicon("./src/apps/http/feed/favicon.ico"));
  koa.use(feedMiddleware);

  return koa;
};
