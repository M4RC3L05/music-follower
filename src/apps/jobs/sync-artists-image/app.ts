import { makeLogger } from "#src/common/logger/mod.ts";
import type { Artist } from "#src/database/types/mod.ts";
import type { CustomDatabase } from "#src/database/mod.ts";
import { delay } from "@std/async";
import { getArtistById } from "#src/remote/itunes/requests/mod.ts";
import { getArtistImage } from "#src/remote/apple-music/requests/mod.ts";

const log = makeLogger("sync-artists-image-job");

const runner = async ({
  db,
  abort,
}: { db: CustomDatabase; abort: AbortSignal }) => {
  if (abort.aborted) {
    return;
  }

  log.info("Begin artist image sync");

  const artists = db.prepare(`
    select *, row_number() over (order by id) as "index", ti."totalItems" as "totalItems"
    from
      artists,
      (select count(id) as "totalItems" from artists) as ti
    order by id;
  `).iter() as IterableIterator<Artist & { index: number; totalItems: number }>;

  for (const artist of artists) {
    const isLastArtist = artist.index >= artist.totalItems;

    if (abort.aborted) {
      break;
    }

    log.info(
      `Sync image from "${artist.name}" at ${artist.index} of ${artist.totalItems}`,
    );

    let imageUrl: string;

    try {
      const artistRemote = await getArtistById(artist.id, abort);

      if (!artistRemote) {
        throw new Error(`No remote artist found with id ${artist.id}`);
      }

      imageUrl = await getArtistImage(
        artistRemote.artistLinkUrl,
        abort,
      );
    } catch (error: unknown) {
      log.error("Something went wrong fetching artist image", { error });

      if (!isLastArtist) {
        log.info("Waiting 5 seconds before processing next artist");

        await delay(5000, { signal: abort }).catch(() => {});
      }

      continue;
    }

    if (abort.aborted) {
      break;
    }

    try {
      log.info("Updating artist image", { id: artist.id });

      db.sql`update artists set image = ${imageUrl} where id = ${artist.id}`;

      log.info("Artist image updated", { id: artist.id });
    } catch (error: unknown) {
      log.error("Something wrong ocurred while updating artist image", {
        error,
      });
    }

    if (!isLastArtist) {
      log.info("Waiting 5 seconds before processing next artist");

      await delay(5000, { signal: abort }).catch(() => {});
    }
  }

  log.info("Artist image sync ended");
};

export default runner;
