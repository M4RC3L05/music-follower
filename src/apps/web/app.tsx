import { serveStatic } from "@hono/hono/deno";
import config from "config";
import { type ContextVariableMap, Hono } from "@hono/hono";
import {
  MemoryStore,
  type Session,
  sessionMiddleware,
} from "@jcs224/hono-sessions";
import { csrf } from "@hono/hono/csrf";
import { jsxRenderer } from "@hono/hono/jsx-renderer";
import { HTTPException } from "@hono/hono/http-exception";
import { secureHeaders } from "@hono/hono/secure-headers";
import { makeLogger } from "#src/common/logger/mod.ts";
import { router } from "#src/apps/web/routes/mod.ts";
import { MainLayout } from "#src/apps/web/views/common/layouts/main.tsx";
import { serviceRegister } from "#src/middlewares/mod.ts";
import type { CustomDatabase } from "#src/database/mod.ts";
import { PagesErrorPage } from "./views/pages/pages/error.tsx";
import type { MiddlewareHandler } from "@hono/hono";
import { STATUS_TEXT, type StatusCode } from "@std/http";
import type {
  SessionFlashFormErrors,
  SessionFlashMessages,
} from "#src/apps/web/types.ts";
import { MultipartParseError } from "@mjackson/multipart-parser";
import { errors } from "@vinejs/vine";
import { feedRoutes } from "#src/apps/web/routes/feed/mod.ts";

const log = makeLogger("admin");

declare module "@hono/hono" {
  interface ContextVariableMap {
    database: CustomDatabase;
    shutdown: AbortSignal;
    session: Session<
      {
        flashMessages?: SessionFlashMessages;
        flashFormErrors?: SessionFlashFormErrors;
      }
    >;
  }

  interface ContextRenderer {
    (
      content: string | Promise<string>,
      props?: { error?: boolean },
    ): Response;
  }
}

export const makeApp = (deps: Partial<ContextVariableMap>) => {
  const app = new Hono();

  app.onError((error, c) => {
    log.error("Something went wrong!", { error });

    if (error instanceof HTTPException) {
      if (error.res) {
        error.res.headers.forEach((value, key) => {
          c.header(key, value);
        });
      }
    }

    if ((error instanceof HTTPException && error.status === 422)) {
      c.get("session").flash("flashMessages", { error: [error.message] });

      if (error.cause) {
        c.get("session").flash(
          "flashFormErrors",
          error.cause as SessionFlashFormErrors,
        );
      }

      return c.redirect(c.req.header("Referer") ?? "/");
    }

    if (error instanceof errors.E_VALIDATION_ERROR) {
      c.get("session").flash("flashMessages", { error: [error.message] });
      c.get("session").flash(
        "flashFormErrors",
        (error.messages as { message: string; field: string }[]).reduce(
          (acc, curr) => {
            const errMessages = acc[curr.field] ?? [];
            errMessages.push(curr.message);
            acc[curr.field] = errMessages;
            return acc;
          },
          {} as Record<string, string[]>,
        ),
      );

      return c.redirect(c.req.header("Referer") ?? "/");
    }

    if (error instanceof MultipartParseError) {
      c.get("session").flash("flashMessages", { error: [error.message] });
      return c.redirect(c.req.header("Referer") ?? "/");
    }

    const errorMessage = error instanceof HTTPException
      ? (error.message.trim().length > 0 ? error.message : undefined) ??
        STATUS_TEXT[error.status as StatusCode] ??
        "Something broke..."
      : "Something broke...";
    c.status(error instanceof HTTPException ? error.status : 500);

    return c.render(<PagesErrorPage errorMessage={errorMessage} />, {
      error: true,
    });
  });

  app.notFound((c) => {
    c.status(404);
    return c.render(<PagesErrorPage errorMessage="Not found" />, {
      error: true,
    });
  });

  app.use("*", async (c, next) => {
    const nocache = () => {
      // This is important so that we always make sure the browser will not cache the previous page
      // so that the requests are always made.
      c.header("cache-control", "no-cache, no-store, must-revalidate");
      c.header("pragma", "no-cache");
      c.header("expires", "0");
    };

    try {
      await next();

      if (
        !(c.req.path.includes("/favicon") || c.req.path.includes("/public") ||
          c.req.path.includes("/deps"))
      ) {
        nocache();
      } else {
        c.header("Cache-Control", `max-age=${60 * 60 * 24}`);
      }
    } catch (error) {
      if (
        !(c.req.path.includes("/favicon") || c.req.path.includes("/public") ||
          c.req.path.includes("/deps"))
      ) {
        nocache();
      } else {
        c.header("Cache-Control", `max-age=${60 * 60 * 24}`);
      }

      throw error;
    }
  });
  app.use("*", serviceRegister(deps));
  app.use(
    "*",
    secureHeaders({ referrerPolicy: "strict-origin-when-cross-origin" }),
  );

  app.route("/feeds", feedRoutes());
  app.get("/favicon.ico", serveStatic({ root: "./src/apps/web/public" }));
  app.get(
    "/public/*",
    serveStatic({
      root: "./src/apps/web/public",
      rewriteRequestPath: (path) => path.replace("/public", ""),
    }),
  );
  app.get(
    "/deps/*",
    serveStatic({
      root: "./node_modules",
      rewriteRequestPath: (path) => path.replace("/deps", ""),
    }),
  );

  app.use(
    "*",
    sessionMiddleware({
      store: new MemoryStore(),
      encryptionKey: "foobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobar",
      expireAfterSeconds: 60 * 15,
      sessionCookieName: "sid",
      cookieOptions: {
        sameSite: "strict",
        path: "/",
        httpOnly: true,
        secure: true,
      },
    }) as unknown as MiddlewareHandler,
  );

  app.use(
    "*",
    jsxRenderer(({ children, error = false }, c) => {
      const flashMessages = c.get("session").get("flashMessages");

      return (
        <MainLayout
          path={c.req.path}
          flashMessages={flashMessages ?? undefined}
          enableNavbar={!error}
        >
          {children}
        </MainLayout>
      );
    }, {
      docType: true,
      stream: true,
    }),
  );

  app.use(
    "*",
    csrf({
      origin: (origin) => new RegExp(config.get("csrf.origin")).test(origin),
    }),
  );

  if (Deno.env.get("ENV") === "test") {
    app.get("/test/flash-messages", (c) => {
      c.get("session").flash("flashMessages", {
        error: ["error", "error2"],
        info: ["info"],
        success: ["success"],
        warning: ["warning"],
      });

      return c.render("");
    });

    app.get("/test/error", (c) => {
      const { type } = c.req.query();

      if (type === "http") {
        throw new HTTPException(400, {
          res: new Response(null, { headers: { "x-foo": "bar" } }),
        });
      }

      if (type === "http422") {
        throw new HTTPException(422, {
          message: "foo",
          cause: { foo: ["yes"] },
        });
      }

      if (type === "valierr") {
        const e = new errors.E_VALIDATION_ERROR([{
          message: "lorem ipsum",
          field: "foo",
        }, {
          message: "lorem lorem",
          field: "bar",
        }]);

        e.message = "foo";

        throw e;
      }

      if (type === "multparterr") {
        throw new MultipartParseError("foo");
      }

      throw new Error("foo");
    });
  }

  return app.route("/", router());
};
