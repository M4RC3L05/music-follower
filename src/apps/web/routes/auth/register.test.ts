import {
  afterAll,
  beforeAll,
  beforeEach,
  describe,
  it,
} from "@std/testing/bdd";
import { type CustomDatabase, makeDatabase } from "#src/database/mod.ts";
import { testDbUtils, testUtils } from "#src/common/utils/mod.ts";
import { makeApp } from "#src/apps/web/app.tsx";
import type { Hono } from "@hono/hono";
import * as testFixtures from "#src/common/test-fixtures/mod.ts";
import { assertEquals } from "@std/assert";
import { MemoryStore } from "@jcs224/hono-sessions";

let app: Hono;
let db: CustomDatabase;
let memStore: MemoryStore;

beforeAll(async () => {
  db = makeDatabase();
  await testDbUtils.runMigrations(db);

  memStore = new MemoryStore();
  app = makeApp({ database: db, sessioStore: memStore });
});

beforeEach(() => {
  db.exec("delete from artists");
  db.exec("delete from accounts");
  // deno-lint-ignore ban-ts-comment
  // @ts-ignore
  memStore.data.clear();
});

afterAll(() => {
  db.close();
});

describe("GET /auth/register", () => {
  describe("snapshots", () => {
    it("should render register page", async (t) => {
      const res = await app.request("/auth/register");
      const dom = await testUtils.getDom(res);

      await testUtils.expectHtmlSnapshot(t, dom, "body");
    });

    it("should render register page with form errors", async (t) => {
      const postRes = await app.request("/auth/register", {
        method: "post",
        headers: { origin: "http://localhost" },
        body: new URLSearchParams({ username: "", password: "" }),
      });

      const res = await app.request("/auth/register", {
        headers: { cookie: postRes.headers.get("set-cookie")! },
      });
      const dom = await testUtils.getDom(res);

      await testUtils.expectHtmlSnapshot(t, dom, "body");
    });
  });

  it("should show the register form", async () => {
    const res = await app.request("/auth/register");
    const dom = await testUtils.getDom(res);

    testUtils.assertSeeInOrder(dom, [
      '<input name="username"',
      '<input name="password"',
      '<input type="submit" value="Register"',
    ], "body main");
  });

  it("should show register form with form errors if inavlid data provided", async () => {
    const postRes = await app.request("/auth/register", {
      method: "post",
      headers: { origin: "http://localhost" },
      body: new URLSearchParams({ username: "", password: "" }),
    });

    const res = await app.request("/auth/register", {
      headers: { cookie: postRes.headers.get("set-cookie")! },
    });
    const dom = await testUtils.getDom(res);

    testUtils.assertSeeInOrder(dom, [
      '<input name="username"',
      '<div id="username-validation-feedback"',
      '<input name="password"',
      '<div id="password-validation-feedback"',
      '<input type="submit" value="Register"',
    ], "body main");
    testUtils.assertSeeTextInOrder(dom, [
      "The username field must be defined",
      "The password field must be defined",
    ], "body main");
  });
});

describe("POST /auth/register", () => {
  it("should flash error if an account already exists", async () => {
    testFixtures.loadAccount(db);

    const res = await app.request("/auth/register", {
      method: "post",
      headers: { origin: "http://localhost" },
      body: new URLSearchParams({ username: "foo", password: "foobarbiz" }),
    });

    assertEquals(res.status, 302);
    assertEquals(res.headers.get("location"), "/auth/login");

    const res2 = await app.request("/auth/login", {
      headers: { cookie: res.headers.get("set-cookie")! },
    });
    const dom = await testUtils.getDom(res2);

    assertEquals(res2.status, 200);
    testUtils.assertSeeText(
      dom,
      "An account already exists",
      "body #flash-messages",
    );
  });

  it("should create a new account", async () => {
    const res = await app.request("/auth/register", {
      method: "post",
      headers: { origin: "http://localhost" },
      body: new URLSearchParams({ username: "foo", password: "foobarbiz" }),
    });

    assertEquals(res.status, 302);
    assertEquals(res.headers.get("location"), "/auth/login");

    const res2 = await app.request("/auth/login", {
      headers: { cookie: res.headers.get("set-cookie")! },
    });
    const dom = await testUtils.getDom(res2);

    assertEquals(res2.status, 200);
    testUtils.assertSeeText(
      dom,
      "Account created, you may login now",
      "body #flash-messages",
    );

    assertEquals(db.sql`select * from accounts`.length, 1);
  });
});
