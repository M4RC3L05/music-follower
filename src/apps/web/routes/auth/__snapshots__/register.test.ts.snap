export const snapshot = {};

snapshot[`global > GET /auth/register > snapshots > should render register page 1`] = `
'
<body>
  <nav class="navbar navbar-expand-lg bg-body-tertiary">
    <div class="container">
      <a
        class="navbar-brand"
        href="/auth/login"
      >
        Music Follower
      </a>
      <button
        class="navbar-toggler"
        type="button"
        data-bs-toggle="collapse"
        data-bs-target="#navbarSupportedContent"
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
      >
        <span class="navbar-toggler-icon">
        </span>
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <ul class="navbar-nav me-auto mb-2 mb-lg-0">
          <li class="nav-item">
            <a
              class="nav-link "
              href="/auth/login"
            >
              Login
            </a>
          </li>
          <li class="nav-item">
            <a
              class="nav-link active"
              href="/auth/register"
            >
              Register
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <main>
    <div class="container mt-5">
      <div class="row">
        <div class="col-sm-8 offset-sm-2">
          <h1 class="text-center">
            Register
          </h1>
          <form
            action="/auth/register"
            method="post"
          >
            <div class="mb-3">
              <label
                for="username"
                class="form-label"
              >
                Username
              </label>
              <input
                name="username"
                id="username"
                class="form-control "
                required
              >
            </div>
            <div class="mb-3">
              <label
                for="password"
                class="form-label"
              >
                Password
              </label>
              <input
                name="password"
                id="password"
                class="form-control "
                required
                type="password"
              >
            </div>
            <div class="d-grid">
              <input
                type="submit"
                value="Register"
                class="btn btn-outline-primary"
              >
            </div>
          </form>
        </div>
      </div>
    </div>
  </main>
  <script
    type="module"
    src="/deps/bootstrap/dist/js/bootstrap.bundle.js"
  >
  </script>
</body>
'
`;

snapshot[`global > GET /auth/register > snapshots > should render register page with form errors 1`] = `
'
<body>
  <nav class="navbar navbar-expand-lg bg-body-tertiary">
    <div class="container">
      <a
        class="navbar-brand"
        href="/auth/login"
      >
        Music Follower
      </a>
      <button
        class="navbar-toggler"
        type="button"
        data-bs-toggle="collapse"
        data-bs-target="#navbarSupportedContent"
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
      >
        <span class="navbar-toggler-icon">
        </span>
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <ul class="navbar-nav me-auto mb-2 mb-lg-0">
          <li class="nav-item">
            <a
              class="nav-link "
              href="/auth/login"
            >
              Login
            </a>
          </li>
          <li class="nav-item">
            <a
              class="nav-link active"
              href="/auth/register"
            >
              Register
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    id="flash-messages"
    class="position-fixed z-3 vstack gap-2 overflow-x-hidden overflow-y-scroll"
    style="bottom:1.3rem;right:1.3rem;width:600px;top:1.3rem;pointer-events:none"
  >
    <div
      class="alert alert-danger alert-dismissible fade show mb-0 d-flex align-items-center"
      role="alert"
      style="pointer-events:all"
    >
      <i class="bi bi-x-circle me-2">
      </i>
      Validation failure
      <button
        type="button"
        class="btn-close"
        data-bs-dismiss="alert"
        aria-label="Close"
      >
      </button>
    </div>
  </div>
  <main>
    <div class="container mt-5">
      <div class="row">
        <div class="col-sm-8 offset-sm-2">
          <h1 class="text-center">
            Register
          </h1>
          <form
            action="/auth/register"
            method="post"
          >
            <div class="mb-3">
              <label
                for="username"
                class="form-label"
              >
                Username
              </label>
              <input
                name="username"
                id="username"
                class="form-control is-invalid"
                required
                aria-describedby="username-validation-feedback"
              >
              <div
                id="username-validation-feedback"
                class="invalid-feedback"
              >
                <p key="0">
                  The username field must be defined
                </p>
              </div>
            </div>
            <div class="mb-3">
              <label
                for="password"
                class="form-label"
              >
                Password
              </label>
              <input
                name="password"
                id="password"
                class="form-control is-invalid"
                required
                type="password"
                aria-describedby="password-validation-feedback"
              >
              <div
                id="password-validation-feedback"
                class="invalid-feedback"
              >
                <p key="0">
                  The password field must be defined
                </p>
              </div>
            </div>
            <div class="d-grid">
              <input
                type="submit"
                value="Register"
                class="btn btn-outline-primary"
              >
            </div>
          </form>
        </div>
      </div>
    </div>
  </main>
  <script
    type="module"
    src="/deps/bootstrap/dist/js/bootstrap.bundle.js"
  >
  </script>
</body>
'
`;
